aws_lambda_powertools/utilities/data_classes/kinesis_firehose_event.py

```diff
@@ -1,9 +1,125 @@
 import base64
-from typing import Iterator, Optional
+import json
+from dataclasses import dataclass, field
+from typing import Any, Callable, Dict, Iterator, List, Optional
+
+from typing_extensions import Literal
 
 from aws_lambda_powertools.utilities.data_classes.common import DictWrapper
 
 
+@dataclass(repr=False, order=False, frozen=True)
+class KinesisFirehoseDataTransformationRecordMetadata:
+    """
+    Documentation:
+    --------------
+    - https://docs.aws.amazon.com/firehose/latest/dev/dynamic-partitioning.html
+    """
+
+    partition_keys: Optional[Dict[str, str]]
+
+    def asdict(self) -> Dict:
+        if self.partition_keys is not None:
+            return {"partitionKeys": self.partition_keys}
+        return {}
```

Comment:
```suggestion
    partition_keys: Dict[str, str] = field(default_factory=lambda: {})
```

Tip: Any Python class by default has a `__dict__` dunder method OR `vars(class_instance)` which dumps the attributes as a dictionary. 

If you were to create a factory with a different structure it'd be fine to have `asdict` like you have for the Response class.

Here, that logic can be fully replaced with a `__dict__`, `vars()`, or even `dataclasses.asdict`. The latter is the slowest but perfect when you have nested dataclasses.

tip 2: have good defaults to avoid checking on them later (mypy, if, etc.) unless you're going to serialize them in a different way OR need them as `None` to assert intentional falsy behaviour. Here, we can use `default_factory()` to instantiate an empty dict during dataclass initialization IF a customer doesn't pass one.

Reply:
For this, we still need `partitionKeys` in dict instead of `partition_keys` that would be produced by `__dict__`